File: trie.py
class node:
    def __init__(self,data):
        self.isTerminal = False
        self.data = data
        self.h = {}

class Trie:
    
    def __init__(self):
        self.root = node(None)
    
    def addWord(self,word):
        temp = self.root
        print(temp)
        for i in range(0,len(word)):
            ch = word[i]
            if ch not in temp.h:
                child = node(ch)
                temp.h[ch] = child
                temp = child
            else:
                temp = temp.h[ch]

        temp.isTerminal = True

    def search(self,word):
        temp = self.root
        for i in range(0,len(word)):
            ch = word[i]
            if ch not in temp.h:
                return False
            else:
                temp = temp.h[ch]
                
        return temp.isTerminal
                        
words = ["apple","ape","coder","satwik"]
global root
t = Trie()

for x in words:
    t.addWord(x)

File: DS and Algos/Trie.cpp
#include<iostream>
#include<unordered_map>

using namespace std;

#define hashmap unordered_map<char,node*>

 class node{
 public:
 	char data;
 	hashmap h;
 	bool isTerminal;

 	node(char d){
 		data = d;
 		isTerminal = false;
 	}
 };

 class Trie{
 	 node* root;
 public:
 	Trie(){
 		root = new node('\0');
 	}

 	//Insertion
 	void addWord(char* word){
 		node* temp = root;
 		for(int i=0;word[i]!='\0';i++){
 			char ch = word[i];
 			if (temp->h.count(ch)==0){
 				node* child = new node(ch);
 				temp->h[ch] = child; //its basically temp->next = child
 				temp = child;
 	
 			else{
 				temp = temp->h[ch];
 			}
 			temp->isTerminal = true;
 		}
 	}

 	bool search(char* word){
 		node* temp = root;
 		for(int i=0;word[i]!='\0';i++){
 			char ch = word[i];
 			if(temp->h.count(ch)!=0){
 				temp=temp->h[ch];
 			}
 			else{
 				return false;
 			}
 		}
 		return temp->isTerminal;
 	}

 }; 

 int main()
 {
 	Trie t;
 	char word[10][100] = {"apple","ape","coder","satwik"};

 	for(int i=0;i<5;i++){
 		t.addWord(word[i]);
 	}

 	char searchword[100]
; 	cin.getline(searchword,100);

 	if(t.search(searchword)){
 		cout<<searchword<<" found"<<endl;
 	}
 	else{
 		cout<<searchword<<" not found"<<endl;
 	}

 	return 0;
 }